File: ontrack-extension-combined/src/main/java/net/nemerosa/ontrack/extension/issues/combined/CombinedIssueServiceExtension.java
package net.nemerosa.ontrack.extension.issues.combined;

import com.google.common.collect.Sets;
import net.nemerosa.ontrack.extension.api.model.IssueChangeLogExportRequest;
import net.nemerosa.ontrack.extension.issues.IssueServiceExtension;
import net.nemerosa.ontrack.extension.issues.IssueServiceRegistry;
import net.nemerosa.ontrack.extension.issues.export.ExportFormat;
import net.nemerosa.ontrack.extension.issues.export.ExportedIssues;
import net.nemerosa.ontrack.extension.issues.model.ConfiguredIssueService;
import net.nemerosa.ontrack.extension.issues.model.Issue;
import net.nemerosa.ontrack.extension.issues.model.IssueServiceConfiguration;
import net.nemerosa.ontrack.extension.support.AbstractExtension;
import net.nemerosa.ontrack.model.support.MessageAnnotation;
import net.nemerosa.ontrack.model.support.MessageAnnotator;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

@Component
public class CombinedIssueServiceExtension extends AbstractExtension implements IssueServiceExtension {

    public static final String SERVICE = "combined";
    private final CombinedIssueServiceConfigurationService configurationService;
    private final IssueServiceRegistry issueServiceRegistry;

    @Autowired
    public CombinedIssueServiceExtension(
            CombinedIssueServiceExtensionFeature extensionFeature,
            IssueServiceRegistry issueServiceRegistry,
            CombinedIssueServiceConfigurationService configurationService) {
        super(extensionFeature);
        this.issueServiceRegistry = issueServiceRegistry;
        this.configurationService = configurationService;
    }

    /**
     * Gets the list of attached configured issue services.
     *
     * @param issueServiceConfiguration Configuration of the combined issue service
     * @return List of associated configured issue services
     */
    protected Collection<ConfiguredIssueService> getConfiguredIssueServices(IssueServiceConfiguration issueServiceConfiguration) {
        CombinedIssueServiceConfiguration combinedIssueServiceConfiguration = (CombinedIssueServiceConfiguration) issueServiceConfiguration;
        return combinedIssueServiceConfiguration.getIssueServiceConfigurationIdentifiers().stream()
                .map(issueServiceRegistry::getConfiguredIssueService)
                .collect(Collectors.toList());
    }

    @Override
    public String getId() {
        return SERVICE;
    }

    @Override
    public String getName() {
        return "Combined issue service";
    }

    @Override
    public List<? extends IssueServiceConfiguration> getConfigurationList() {
        return configurationService.getConfigurationList();
    }

    @Override
    public IssueServiceConfiguration getConfigurationByName(String name) {
        return configurationService.getConfigurationByName(name).orElse(null);
    }

    /**
     * Without any specific configuration, we have to assume the token is valid.
     */
    @Override
    public boolean validIssueToken(String token) {
        return true;
    }

    @Override
    public String getIssueExtractionRegex(IssueServiceConfiguration issueServiceConfiguration) {
        return getConfiguredIssueServices(issueServiceConfiguration).stream()
                .map(ConfiguredIssueService::getIssueExtractionRegex)
                .collect(Collectors.joining("|"));
    }

    @Override
    public Set<String> extractIssueKeysFromMessage(IssueServiceConfiguration issueServiceConfiguration, String message) {
        return getConfiguredIssueServices(issueServiceConfiguration).stream()
                .map(
                        configuredIssueService ->
                                configuredIssueService.getIssueServiceExtension().extractIssueKeysFromMessage(
                                        configuredIssueService.getIssueServiceConfiguration(),
                                        message
                                )
                )
                .collect(
                        Collectors.reducing(
                                Collections.<String>emptySet(),
                                Sets::union
                        )
                );
    }

    @Override
    public Optional<MessageAnnotator> getMessageAnnotator(IssueServiceConfiguration issueServiceConfiguration) {
        // Gets all the defined message annotators
        Collection<MessageAnnotator> messageAnnotators = getConfiguredIssueServices(issueServiceConfiguration).stream()
                .map(
                        configuredIssueService ->
                                configuredIssueService.getIssueServiceExtension().getMessageAnnotator(
                                        configuredIssueService.getIssueServiceConfiguration()
                                )
                )
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.toList());
        if (messageAnnotators.isEmpty()) {
            return Optional.empty();
        } else {
            // For each message annotator, gets the list of annotation
            return Optional.of(
                    text -> messageAnnotators.stream()
                            .map(messageAnnotator -> messageAnnotator.annotate(text))
                            .map((Function<Collection<MessageAnnotation>, HashSet<MessageAnnotation>>) HashSet::new)
                            .collect(
                                    // ... and gets them all together
                                    Collectors.reducing(
                                            Collections.<MessageAnnotation>emptySet(),
                                            Sets::union
                                    )
                            )
            );
        }
    }

    @Override
    public String getLinkForAllIssues(IssueServiceConfiguration issueServiceConfiguration, List<Issue> issues) {
        return null;
    }

    @Override
    public Issue getIssue(IssueServiceConfiguration issueServiceConfiguration, String issueKey) {
        return getConfiguredIssueServices(issueServiceConfiguration).stream()
                .map(
                        configuredIssueService ->
                                configuredIssueService.getIssueServiceExtension().getIssue(
                                        configuredIssueService.getIssueServiceConfiguration(),
                                        issueKey
                                )
                )
                .filter(issue -> issue != null)
                .findFirst()
                .orElse(null)
                ;
    }

    @Override
    public boolean containsIssueKey(IssueServiceConfiguration issueServiceConfiguration, String key, Set<String> keys) {
        return getConfiguredIssueServices(issueServiceConfiguration).stream()
                .anyMatch(
                        configuredIssueService ->
                                configuredIssueService.getIssueServiceExtension().containsIssueKey(
                                        configuredIssueService.getIssueServiceConfiguration(),
                                        key,
                                        keys
                                )
                );
    }

    @Override
    public List<ExportFormat> exportFormats(IssueServiceConfiguration issueServiceConfiguration) {
        Set<ExportFormat> lists = getConfiguredIssueServices(issueServiceConfiguration).stream()
                .map(
                        configuredIssueService ->
                                configuredIssueService.getIssueServiceExtension().exportFormats(
                                        configuredIssueService.getIssueServiceConfiguration()
                                )
                )
                .map((Function<List<ExportFormat>, HashSet<ExportFormat>>) HashSet::new)
                .collect(
                        // ... and gets them all together
                        Collectors.reducing(
                                Collections.<ExportFormat>emptySet(),
                                Sets::union
                        )
                );
        return new ArrayList<>(lists);
    }

    @Override
    public ExportedIssues exportIssues(IssueServiceConfiguration issueServiceConfiguration, List<? extends Issue> issues, IssueChangeLogExportRequest request) {
        List<ExportedIssues> exportedIssues = getConfiguredIssueServices(issueServiceConfiguration).stream()
                .map(
                        configuredIssueService ->
                                configuredIssueService.getIssueServiceExtension().exportIssues(
                                        configuredIssueService.getIssueServiceConfiguration(),
                                        issues,
                                        request
                                )
                )
                .collect(Collectors.toList());
        // Checks the format is the same for all exports (it must)
        if (!exportedIssues.stream().allMatch(it -> StringUtils.equals(it.getFormat(), request.getFormat()))) {
            throw new IllegalStateException("All exported issues must have the same export format");
        }
        // Concatenates the content
        return new ExportedIssues(
                request.getFormat(),
                exportedIssues.stream().map(ExportedIssues::getContent).collect(Collectors.joining(""))
        );
    }

    @Override
    public Optional<String> getIssueId(IssueServiceConfiguration issueServiceConfiguration, String token) {
        return getConfiguredIssueServices(issueServiceConfiguration).stream()
                .map(
                        configuredIssueService ->
                                configuredIssueService.getIssueServiceExtension().getIssueId(
                                        configuredIssueService.getIssueServiceConfiguration(),
                                        token
                                )
                )
                .filter(Optional::isPresent)
                .findFirst()
                .orElse(Optional.<String>empty());
    }
}

File: ontrack-extension-combined/src/main/java/net/nemerosa/ontrack/extension/issues/combined/CombinedIssueServiceExtension.kt
package net.nemerosa.ontrack.extension.issues.combined

import net.nemerosa.ontrack.extension.api.model.IssueChangeLogExportRequest
import net.nemerosa.ontrack.extension.issues.IssueServiceExtension
import net.nemerosa.ontrack.extension.issues.IssueServiceRegistry
import net.nemerosa.ontrack.extension.issues.export.ExportFormat
import net.nemerosa.ontrack.extension.issues.export.ExportedIssues
import net.nemerosa.ontrack.extension.issues.model.ConfiguredIssueService
import net.nemerosa.ontrack.extension.issues.model.Issue
import net.nemerosa.ontrack.extension.issues.model.IssueServiceConfiguration
import net.nemerosa.ontrack.extension.support.AbstractExtension
import net.nemerosa.ontrack.model.support.MessageAnnotator
import org.apache.commons.lang3.StringUtils
import org.springframework.stereotype.Component
import java.util.*
import kotlin.jvm.optionals.getOrNull

@Component
class CombinedIssueServiceExtension(
    extensionFeature: CombinedIssueServiceExtensionFeature,
    private val issueServiceRegistry: IssueServiceRegistry,
    private val configurationService: CombinedIssueServiceConfigurationService
) : AbstractExtension(extensionFeature), IssueServiceExtension {

    /**
     * Gets the list of attached configured issue services.
     *
     * @param issueServiceConfiguration Configuration of the combined issue service
     * @return List of associated configured issue services
     */
    protected fun getConfiguredIssueServices(issueServiceConfiguration: IssueServiceConfiguration): Collection<ConfiguredIssueService> {
        val combinedIssueServiceConfiguration = issueServiceConfiguration as CombinedIssueServiceConfiguration
        return combinedIssueServiceConfiguration
            .issueServiceConfigurationIdentifiers
            .mapNotNull { issueServiceRegistry.getConfiguredIssueService(it) }
    }

    override fun getId(): String {
        return SERVICE
    }

    override fun getName(): String {
        return "Combined issue service"
    }

    override fun getConfigurationList(): List<IssueServiceConfiguration> {
        return configurationService.configurationList
    }

    override fun getConfigurationByName(name: String): IssueServiceConfiguration {
        return configurationService.getConfigurationByName(name).orElse(null)
    }

    override fun getMessageRegex(issueServiceConfiguration: IssueServiceConfiguration, issue: Issue): String {
        return getConfiguredIssueServices(issueServiceConfiguration)
            .firstOrNull()
            ?.getMessageRegex(issue)
            ?: ""
    }

    override fun extractIssueKeysFromMessage(
        issueServiceConfiguration: IssueServiceConfiguration,
        message: String
    ): Set<String> {
        return getConfiguredIssueServices(issueServiceConfiguration)
            .map { configuredIssueService ->
                configuredIssueService.issueServiceExtension.extractIssueKeysFromMessage(
                    configuredIssueService.issueServiceConfiguration,
                    message
                )
            }
            .fold(emptySet()) { acc, values ->
                acc + values
            }
    }

    override fun getMessageAnnotator(issueServiceConfiguration: IssueServiceConfiguration): MessageAnnotator? {
        // Gets all the defined message annotators
        val messageAnnotators = getConfiguredIssueServices(issueServiceConfiguration)
            .mapNotNull { configuredIssueService ->
                configuredIssueService.issueServiceExtension.getMessageAnnotator(
                    configuredIssueService.issueServiceConfiguration
                )
            }
        return if (messageAnnotators.isEmpty()) {
            null
        } else {
            // For each message annotator, gets the list of annotation
            val messageAnnotator = MessageAnnotator { text ->
                messageAnnotators
                    .map { messageAnnotator -> messageAnnotator.annotate(text).toSet() }
                    .fold(emptySet()) { acc, values -> acc + values }
            }
            messageAnnotator
        }
    }

    override fun getIssue(issueServiceConfiguration: IssueServiceConfiguration, issueKey: String): Issue? {
        return getConfiguredIssueServices(issueServiceConfiguration).firstNotNullOfOrNull { configuredIssueService ->
            configuredIssueService.issueServiceExtension.getIssue(
                configuredIssueService.issueServiceConfiguration,
                issueKey
            )
        }
    }

    @Deprecated("Deprecated in Java")
    override fun exportFormats(issueServiceConfiguration: IssueServiceConfiguration): List<ExportFormat> {
        return getConfiguredIssueServices(issueServiceConfiguration)
            .map { configuredIssueService ->
                configuredIssueService.issueServiceExtension.exportFormats(
                    configuredIssueService.issueServiceConfiguration
                ).toSet()
            }
            .fold(emptySet<ExportFormat>()) { acc, v -> acc + v }
            .toList()
    }

    @Deprecated("Deprecated in Java")
    override fun exportIssues(
        issueServiceConfiguration: IssueServiceConfiguration,
        issues: List<Issue>,
        request: IssueChangeLogExportRequest
    ): ExportedIssues {
        val exportedIssues = getConfiguredIssueServices(issueServiceConfiguration)
            .map { configuredIssueService ->
                configuredIssueService.issueServiceExtension.exportIssues(
                    configuredIssueService.issueServiceConfiguration,
                    issues,
                    request
                )
            }
        // Checks the format is the same for all exports (it must)
        check(exportedIssues.all { it ->
            StringUtils.equals(
                it.format,
                request.format
            )
        }) { "All exported issues must have the same export format" }
        // Concatenates the content
        return ExportedIssues(
            request.format,
            exportedIssues.mapNotNull { exportedIssue ->
                exportedIssue.content.takeIf { it.isNotBlank() }
            }.joinToString("\n")
        )
    }

    override fun getIssueId(issueServiceConfiguration: IssueServiceConfiguration, token: String): String? {
        return getConfiguredIssueServices(issueServiceConfiguration)
            .firstNotNullOfOrNull { configuredIssueService ->
                configuredIssueService.issueServiceExtension.getIssueId(
                    configuredIssueService.issueServiceConfiguration,
                    token
                )
            }
    }

    override fun getIssueTypes(
        issueServiceConfiguration: IssueServiceConfiguration,
        issue: Issue,
    ): Set<String> =
        getConfiguredIssueServices(issueServiceConfiguration).map {
            it.issueServiceExtension.getIssueTypes(it.issueServiceConfiguration, issue)
        }.reduce { acc, set ->
            acc + set
        }

    companion object {
        const val SERVICE = "combined"
    }
}
